docs/shared/recipes/running-tasks/customizing-inputs.md

```diff
@@ -44,136 +48,282 @@ To understand some of the examples below, let's imagine the following simple wor
+
+For example, running `nx build myreactapp --prod` will not reuse the same results of running `nx build myreactapp` earlier.
+
+### Source Files
+
+Changing source code will often result in a change of behavior of running different tasks. Nx can consider the contents of files matching a pattern when calculating the computation hash.
+Source File inputs are defined as such:
+
+```jsonc
+"inputs": {
+  "{projectRoot}/**/*", // All files in a project
+  "{workspaceRoot}/.gitignore", // A specific file in the workspace
+  "{projectRoot}/**/*.ts", // A glob pattern for files
+  "!{projectRoot}/**/*.spec.ts", // Excluding files matching a glob pattern
+}
+```
+
+Source File inputs must be prefixed with either `{projectRoot}` or `{workspaceRoot}` to distinguish where the paths should be resolved from.
+
+Prefixing a Source File input with `!` will exclude the files matching the pattern from the set of files used to calculate the hash.
+
+By default, Nx will use all files in a project as well as all files in the project's dependencies when computing a hash for tasks belonging to the project.
+This may cause Nx to rerun some tasks even when files irrelevant to the task have changed but it ensures that by default, Nx always re-runs the task when it should.
+
+See the [Common Input Sets section](#common-inputs) to see some common examples of source file inputs
+
+### Environment Variables
+
+Tools and scripts will often use some Environment Variables to change their behavior. Nx can hash the value of environment variables when it runs in the cache hash key calculation in order to invalidate the cache if the environment variable value changes. Environment Variable inputs are defined as such:
+
+```jsonc
+"inputs": [
+  { "env": "API_KEY" } // this will include the value of $API_KEY in the cache hash
+]
+```
+
+### Runtime Inputs
+
+You can use a Runtime input to provide a script which will output the information you want to include in the computation hash. Runtime inputs are defined as such:
+
+```jsonc
+"inputs": [
+  { "runtime": "node --version" }
+]
+```
+
+This kind of input is often used to include versions of tools used to run the task. An advanced use case for this type of input is to provide a custom script which will have its output considered in the hash.
+
+### External Dependencies
+
+Source code often imports from external dependencies from installed through package managers. For example, a React application will likely import `react`. Nx will always consider External Dependencies depended upon by any source code within the project.
+
+External Dependencies are also often tools used to run tasks. Updating the versions of those tools will change the behavior of those tasks. External Dependencies inputs are defined as such:
```

Comment:
We should make it clearer that the external dependencies are only effective if you're running a task without an executor, i.e., just the command option or with a plugin that implicitly creates targets. It does not affect the dependencies of the actual project